### ISSUE TITLE
Show loading while buffering 

### ISSUE THREAD

**Issue by AmirHadifar** [brianwernick/ExoMedia#102]
In your next release you can consider this it would be nice in low-internet speed connections.

**Comment by brianwernick** [comment]
I'm glad you brought this up, I'll try to get it in to the 3.0 release

**Comment by aarongrider** [comment]
I think I just came across a bug with the `OnBufferingUpdateListener()` not getting called from an `EMAudioPlayer` instance that uses `ExoPlayer` even when we have set the listener and implemented it. I'm not sure if this has to do with this bug specifically, but I can't find a way to get callbacks from `EMAudioPlayer` on buffering updates.

**Comment by brianwernick** [comment]
@aarongrider, The ExoPlayer doesn't inform us of Buffering updates which is why the `OnBufferingUpdateListener` is not called, nor is it related because this is a feature request not a bug report

**Comment by kashifandroiddev** [comment]
does 3.0 preview have this feature of show buffring while loading i want to test it ?

**Comment by brianwernick** [comment]
@kashifandroiddev no, this hasn't been added yet (otherwise the request would have been closed)

**Comment by appyo** [comment]
You can add new ExoPlayerListener to your Videoview and in onStateChanged it fires "Exoplayer.STATE_BUFFERING" as playbackstate when it's buffering. I tested with exomedia 2.5.6 and  it works.

**Comment by brianwernick** [comment]
@appyo While you can register that listener it doesn't provide quite the functionality this addresses.  The `STATE_BUFFERING` represents a fairly short cache of the video, on slow internet connections this may only be a second or so, which will play then hit the buffering again.  Ideally we would monitor the network speed and determine a minimum amount (or allow it to be specified) that we need buffered before we automatically resume playback.

What you mentioned is already used (in 3.0) to represent loading when seeking and initial load

**Comment by appyo** [comment]
Thanks @brianwernick for your information. I did not know that and i use this only for showing progressbar or similar to inform the user that the video freezes briefly and he does not think that the player has crashed or so. You are right of course it is better if you could adjust the buffering to the network speed. I will test the 3.0 soon, many thanks for your great work.

**Comment by miladz68** [comment]
this feature is a great. I had implemented this in 2.5.6 by adding an exoplayerlistener which is not available in version 3.0.0

**Comment by vishnuganta22** [comment]
how can i know that video is buffering in version 3.0.1

**Comment by IOException722** [comment]
hey @brianwernick 

Thanks for the great library.

Issue is still opened snce 2 year. Should we wait for issue to be fixed or try different options ?

Do let me know as we are moving into production.

Thanks

**Comment by brianwernick** [comment]
I'm still planning on adding the functionality; it's actually one of the main tasks for the 5.0 release. However I don't know how soon I'll be able to get to it (I'm hoping in the next month)

**Comment by deebash** [comment]
Have you added that feature?

**Comment by brianwernick** [comment]
There are two parts to this request:
1. Improving the existing functionality to show loading while buffering when playback catches up to the buffer. 
2. Adding the capabilities to track the buffer / download speed and dynamically adjust the playback resumption.

With the work that we have already done in 5.0 to improve state communication has addressed part 1 above. Because part 1 seems to be the more important part of the request for most I'm going to move part 2 to a future release so that we can get the 5.0 release out sooner
